### ISSUE TITLE
Feature Request: Early Stopping with the tf.estimator

### ISSUE THREAD

**Issue by ankitvgupta** [tensorflow/tensorflow#18394]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: Pip
- **TensorFlow version (use command below)**: 1.7
- **Python version**: 3.6
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: 9
- **GPU model and memory**: NVIDIA K80
- **Exact command to reproduce**: N/A

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

I would love for there to be a SessionRunHook implementation in tensorflow that implements early stopping. Now that `tf.contrib.learn` is being officially deprecated, the existing way that I did early stopping (using a ValidationMonitor) is no longer an option. This seems like a super important feature to have.

The docs indicate in several places that you can simply extend a SessionRunHook to do this, and that seems reasonable. I think that having a standard way to do this would be super useful to lots of users, perhaps even directly built into TrainSpec.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

N/A

**Comment by martinwicke** [comment]
@ispirmustafa @salehay 

We have an early stopping hook somewhere, no? Can we move it to TensorFlow?

**Comment by salehay** [comment]
Yes we have one in TFX, @ispirmustafa let's chat about moving to core.

**Comment by ankitvgupta** [comment]
Great, thanks! Do you guys have any approximate timeline for when this could happen? Just trying to decide whether to hack something together ourselves in the meantime.

**Comment by zmjjmz** [comment]
My understanding at the moment (from having tried to implement this as a SessionRunHook) is that when using `train_and_evaluate`, which I believe is the canonical way to use an Estimator (and is how SageMaker does it), you cannot use a `run_context.request_stop()` call to actually stop training.

I know this is a larger issue with how `train_and_evaluate` works, but are there plans to be able to use any sort of Early Stopping hook w/in that framework?

Thanks!

**Comment by ankitvgupta** [comment]
Interesting, is there any reason that we can’t use a similar logic as StopAtStepHook? Where, rather than checking the step number every time, it runs validation every n steps and then uses the same stop mechanism?

**Comment by martinwicke** [comment]
That works for the single worker case, but with multiple workers, all of them need to know to stop running. We have an implementation of this, basically, we have the eval worker check if it wants to stop after running eval. If it wants to stop, it sets a REQUEST_STOP variable on the parameter server. The other workers run a Hook which checks this variables and stop if a stop has been requested.

**Comment by ankitvgupta** [comment]
Ok great, would be great to get that merged!

**Comment by pmitche** [comment]
Seconded, this feature would be a great addition.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ankitvgupta** [comment]
@salehay @martinwicke just wanted to check in on this - any update on getting this from TFX into tensorflow?

**Comment by ispirmustafa** [comment]
@xiejw could you please take a look?

**Comment by ankitvgupta** [comment]
It looks like in the `tf.estimator.train_and_evaluate` code, there is a line at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/training.py#L432
where we instantiate a `_TrainingExecutor`:

``` python
executor = _TrainingExecutor(estimator=estimator, train_spec=train_spec, eval_spec=eval_spec)
```
However, `_TrainingExecutor` also has the optional argument `continuous_eval_listener`, which we could presumably use to track the eval result and do early stopping. Is there a reason tf.train_and_evaluate doesn't let us use this argument? It doesn't look like there are any other instances of _TrainingExecutor anywhere in the tensorflow codebase.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @xiejw, @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ispirmustafa** [comment]
@goutham kindly volunteered on providing TFX early stopping utils in TensorFlow.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @goutham, @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ankitvgupta** [comment]
Any update on this @goutham?

**Comment by goutham** [comment]
@ankitvgupta Under code review internally. ETA: < 1 week from now.

**Comment by ankitvgupta** [comment]
Fantastic! I appreciate the update @goutham.

**Comment by carlthome** [comment]
Finally! 🎉

Early stopping is a crucial component in a proper DL library and something I sorely miss from Keras. Having this in `tf.estimator` would greatly simplify https://github.com/tensorflow/tensorflow/issues/16576 and put the final nail in the coffin regarding https://stackoverflow.com/questions/47137061/early-stopping-with-tf-estimator-how

**Comment by Marko-Stamenovic-Bose** [comment]
Excited about this too! We have been using home-brewed versions of early stopping with combinations of hooks and environmental variables but they have proven brittle to changes in the estimator API. A more robust solution is very welcome.

**Comment by goutham** [comment]
It's in: 90e66f2

**Comment by ankitvgupta** [comment]
Awesome, thanks @goutham and team!

**Comment by ankitvgupta** [comment]
Should we expect this to be included in the final release of TF 1.9?

**Comment by martinwicke** [comment]
No. The cutoff for 1.9 is long past. It will be in 1.10.

**Comment by ankitvgupta** [comment]
Makes sense, thanks.

**Comment by ankitvgupta** [comment]
Also to clarify, you mention that this implementation sets a variable to request a stop, and then other workers check that variable. I can see where the PR sets a variable called STOP. 

Where is the logic for other variables to check that the STOP variable is set? Is that being separately merged?

**Comment by goutham** [comment]
@ankitvgupta It's part of the same PR. If you look at [early_stopping.py](https://github.com/tensorflow/tensorflow/blob/90e66f2aa1015496c8f8e9be573ef83f542a2ad0/tensorflow/contrib/estimator/python/estimator/early_stopping.py) implementation, there are two hooks `_StopOnPredicateHook` and `_CheckForStoppingHook`. The former is used in the chief and sets the variable when early stopping condition is met, while the latter on workers and checks if the variable is set. Did you notice any problems?

**Comment by ankitvgupta** [comment]
Ah yes, that makes sense. I was testing this on a single GPU a local training configuration. If I'm only using the local case, should I manually add a second hook which is the `_CheckForStoppingHook`?

**Comment by ankitvgupta** [comment]
I think that the run_local case of train_and_evaluate has to be modified for that to work. Here: https://github.com/tensorflow/tensorflow/blob/r1.9/tensorflow/python/estimator/training.py#L634

**Comment by goutham** [comment]
I don't think you should have any problems running in local mode. It is fine to not have `_CheckForStoppingHook` as it is to help with the distributed case. Can you elaborate on what you noticed? I couldn't reproduce any problems when running locally on my machine.

**Comment by ankitvgupta** [comment]
Interesting, when I run locally, it prints the “requesting stop” message, runs eval/export but then train_and_evaluate just continues training. I assumed this is because of the “while True” in the run_local method.

To be clear, I’m using TF 1.9, and copied that early stopping hook code directly (since it’s not releasing until 1.10).

**Comment by ankitvgupta** [comment]
As, to clarify, this works fine when I use the `train()` method of an estimator, but does not work when I use `tf.estimator.train_and_evaluate()`

**Comment by goutham** [comment]
run_local was also updated to work well with train_and_evaluate. See 3edb609. This should be in 1.10. So, if you really want to use early stopping before 1.10 release, I suggest also patching in this PR. But I recommend waiting for 1.10 so that you don't have more trouble.

**Comment by ankitvgupta** [comment]
Makes sense! I'll wait for 1.10.

**Comment by ankitvgupta** [comment]
Hi @goutham, I tried this out on tf-nightly-gpu, and ran into one small bug. 

I used `tf.train_and_evaluate`, and added the new early stopping hook as one of the train hooks. It looks like it checks for early stopping right after the first checkpoint (step 0), and as a result finds that that is no eval directory made yet since the first eval hasn't happened yet. Specifically, I get this error message (replacing MY_MODEL_DIR with the actual directory)

```tensorflow.python.framework.errors_impl.NotFoundError: $MY_MODEL_DIR/eval; No such file or directory```

In the new version of `run_local`, it looks like evaluation occurs [via a CheckpointSaverListener](https://github.com/tensorflow/tensorflow/blob/3edb609926f2521c726737fc1efeae1572dc6581/tensorflow/python/estimator/training.py#L685), so eval should occur right after the checkpoint is saved. However, it looks like the early stopping hook is running before that happens. 

I believe there is a min_steps parameter, but it appears to be checking for eval stats even if I set that to be high.

**Comment by goutham** [comment]
@ankitvgupta I cannot repro this behavior on my linux workstation, unfortunately. Which OS do you use? Can you provide a fuller stacktrace (the part only tracing through TF code)? That would help.

We access the eval directory in the function `_summaries` [here](https://github.com/tensorflow/tensorflow/blob/6c528feaf820bdde820833ad24e05167adb5daa7/tensorflow/contrib/estimator/python/estimator/early_stopping.py#L397). If the directory does not exist, `gfile.Glob` returns an empty list, so no error should be raised.

**Comment by ankitvgupta** [comment]
Yep that’s what I did too @carlthome.

I’ll try to send a minimal example of this for the TF team.

**Comment by goutham** [comment]
I discovered that in some environments, `gfile.Glob` behaves differently from python [`glob.glob`](https://docs.python.org/2/library/glob.html). This is definitely something to be fixed in `gfile.Glob` but for now I will send out a simple fix in early_stopping library to resolve this.

**Comment by goutham** [comment]
Fix is in: 69f229a
(this has missed the boat on 1.10 though)

**Comment by ankitvgupta** [comment]
Appreciate the help!

**Comment by ankitvgupta** [comment]
You'll want to use train_and_evaluate, so that eval runs every once in a while. The hook just reads the eval metrics and stops when the condition is met. So, if you're not running eval, it'll not stop. Relevant line: https://github.com/tensorflow/tensorflow/blob/4dcfddc5d12018a5a0fdca652b9221ed95e9eb23/tensorflow/contrib/estimator/python/estimator/early_stopping.py#L361

**Comment by mihi-r** [comment]
@ankitvgupta Ah, that makes sense. Thanks!

**Comment by Carl-Jensen-Bose** [comment]
I've had a problem with the the `stop_if_no_decrease` hook and I think I've found the problem in the code I changed below. 

The problem is that the python `dict.items()` method doesn't return entries in the order they went in. The indeterminate ordering means that the hook can trigger when it comes across any eval checkpoints that are more than max_steps apart without having come across a better eval inbetween. 

To fix it, I keep the best eval as well as the most current and check how far apart they are.

```diff
diff --git a/tensorflow/contrib/estimator/python/estimator/early_stopping.py b/tensorflow/contrib/estimator/python/estimator/early_stopping.py
index af4855e..194422b 100644
--- a/tensorflow/contrib/estimator/python/estimator/early_stopping.py
+++ b/tensorflow/contrib/estimator/python/estimator/early_stopping.py
@@ -362,6 +362,8 @@ def _stop_if_no_metric_improvement_hook(

     best_val = None
     best_val_step = None
+    current_val = None
+    current_step = None
     for step, metrics in eval_results.items():
       if step < min_steps:
         continue
@@ -369,13 +371,16 @@ def _stop_if_no_metric_improvement_hook(
       if best_val is None or is_lhs_better(val, best_val):
         best_val = val
         best_val_step = step
+      if current_val is None or operator.gt(step, current_step):
+        current_val = val
+        current_val_step = step
+    if current_step - best_val_step >= max_steps_without_improvement:
+      tf_logging.info(
+          'No %s in metric "%s" for %s steps, which is greater than or equal '
+          'to max steps (%s) configured for early stopping.',
+          increase_or_decrease, metric_name, step - best_val_step,
+          max_steps_without_improvement)
+      return True
-      if step - best_val_step >= max_steps_without_improvement:
-        tf_logging.info(
-            'No %s in metric "%s" for %s steps, which is greater than or equal '
-            'to max steps (%s) configured for early stopping.',
-            increase_or_decrease, metric_name, step - best_val_step,
-            max_steps_without_improvement)
-        return True
     return False

```

**Comment by goutham** [comment]
@Carl-Jensen-Bose Thanks for pointing it out. `read_eval_metrics` should have returned an `OrderedDict` (sorted by global_step key) instead of `dict`. Have a fix out for review.

**Comment by Carl-Jensen-Bose** [comment]
@goutham Even better. Thanks!

**Comment by goutham** [comment]
Fixed in b9e6bbc.

**Comment by ed-alertedh** [comment]
SInce it looks like one or two of the bug fixes mentioned here didn't make it into the 1.11 release I've patched the latest version of `early_stopping.py` so that it can be used standalone (this version uses the public API rather than the internal module names) https://gist.github.com/ed-alertedh/2ea554d9a17a67c87f00cbd617b1ff32

I find this easier than patching the installed package and less risky than switching to a nightly build. Seems to be working in TF 1.10, presumably should be OK with 1.11 too.

**Comment by aneergaard** [comment]
@goutham Is this fix out in 1.12?

**Comment by Apollo2Mars** [comment]
mark

**Comment by plaffitte** [comment]
Does anyone know if this has been taken care of in tf 2 ?

**Comment by martinwicke** [comment]
@plaffitte what are you referring to? I don't think there are open issues (at least not from this thread).

**Comment by plaffitte** [comment]
@martinwicke Sorry for not being clear, as far as I am aware the only option out there to achieve early stopping in tf2 is by using `tf.estimator.experimental.stop_if_no_decrease_hook` but I was wondering if this is stable and working and if there was another solution to this.

**Comment by martinwicke** [comment]
@karmel to keep me honest.

It should work and it is stable in a sense that we're not planning to
remove it or change it. You can also always write your own hook with custom
behavior.

**Comment by plaffitte** [comment]
@martinwicke Ok I understand, the reason I was asking is because it doesn't to be doing anything in my case and since it's in the experimental module I was wondering if it was working at all.
Also, in general, I find it very weird that tensorflow does not have a built-in early stopping mechanism, or something more accessible as a class in the main package.

**Comment by bhack** [comment]
As [much of the efforts](https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/-kdC7VVkyco) are put over Keras as the highlevel API you can alternatively use https://www.tensorflow.org/versions/r2.2/api_docs/python/tf/keras/callbacks/EarlyStopping

**Comment by plaffitte** [comment]
@bhack Yes I understand that Keras is now the preferred high-level API, but I am getting acquainted with Kubeflow and going over the tutorial which all use the Estimator API, which is why I need to stick with it for the moment

**Comment by bhack** [comment]
@plaffitte Yes I know but stuffs are getting in sync:
https://github.com/kubeflow/examples/issues/774
https://github.com/kubeflow/examples/issues/773

**Comment by plaffitte** [comment]
@bhack Ah awesome thank you!
